File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/trajectorysequence/TrajectorySequenceRunner.java
package org.firstinspires.ftc.teamcode.trajectorysequence;

import androidx.annotation.Nullable;

import com.acmerobotics.dashboard.FtcDashboard;
import com.acmerobotics.dashboard.canvas.Canvas;
import com.acmerobotics.dashboard.config.Config;
import com.acmerobotics.dashboard.telemetry.TelemetryPacket;
import com.acmerobotics.roadrunner.control.PIDCoefficients;
import com.acmerobotics.roadrunner.control.PIDFController;
import com.acmerobotics.roadrunner.drive.DriveSignal;
import com.acmerobotics.roadrunner.followers.TrajectoryFollower;
import com.acmerobotics.roadrunner.geometry.Pose2d;
import com.acmerobotics.roadrunner.profile.MotionState;
import com.acmerobotics.roadrunner.trajectory.Trajectory;
import com.acmerobotics.roadrunner.trajectory.TrajectoryMarker;
import com.acmerobotics.roadrunner.util.NanoClock;
import com.qualcomm.robotcore.hardware.VoltageSensor;

import org.firstinspires.ftc.teamcode.trajectorysequence.sequencesegment.SequenceSegment;
import org.firstinspires.ftc.teamcode.trajectorysequence.sequencesegment.TrajectorySegment;
import org.firstinspires.ftc.teamcode.trajectorysequence.sequencesegment.TurnSegment;
import org.firstinspires.ftc.teamcode.trajectorysequence.sequencesegment.WaitSegment;
import org.firstinspires.ftc.teamcode.util.DashboardUtil;
import org.firstinspires.ftc.teamcode.util.LogFiles;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

@Config
public class TrajectorySequenceRunner {
    public static String COLOR_INACTIVE_TRAJECTORY = "#4caf507a";
    public static String COLOR_INACTIVE_TURN = "#7c4dff7a";
    public static String COLOR_INACTIVE_WAIT = "#dd2c007a";

    public static String COLOR_ACTIVE_TRAJECTORY = "#4CAF50";
    public static String COLOR_ACTIVE_TURN = "#7c4dff";
    public static String COLOR_ACTIVE_WAIT = "#dd2c00";

    public static int POSE_HISTORY_LIMIT = 100;

    private final TrajectoryFollower follower;

    private final PIDFController turnController;

    private final NanoClock clock;

    private TrajectorySequence currentTrajectorySequence;
    private double currentSegmentStartTime;
    private int currentSegmentIndex;
    private int lastSegmentIndex;

    private Pose2d lastPoseError = new Pose2d();

    List<TrajectoryMarker> remainingMarkers = new ArrayList<>();

    private final FtcDashboard dashboard;
    private final LinkedList<Pose2d> poseHistory = new LinkedList<>();

    private VoltageSensor voltageSensor;

    private List<Integer> lastDriveEncPositions, lastDriveEncVels, lastTrackingEncPositions, lastTrackingEncVels;

    public TrajectorySequenceRunner(
            TrajectoryFollower follower, PIDCoefficients headingPIDCoefficients, VoltageSensor voltageSensor,
            List<Integer> lastDriveEncPositions, List<Integer> lastDriveEncVels, List<Integer> lastTrackingEncPositions, List<Integer> lastTrackingEncVels
    ) {
        this.follower = follower;

        turnController = new PIDFController(headingPIDCoefficients);
        turnController.setInputBounds(0, 2 * Math.PI);

        this.voltageSensor = voltageSensor;

        this.lastDriveEncPositions = lastDriveEncPositions;
        this.lastDriveEncVels = lastDriveEncVels;
        this.lastTrackingEncPositions = lastTrackingEncPositions;
        this.lastTrackingEncVels = lastTrackingEncVels;

        clock = NanoClock.system();

        dashboard = FtcDashboard.getInstance();
        dashboard.setTelemetryTransmissionInterval(25);
    }

    public void followTrajectorySequenceAsync(TrajectorySequence trajectorySequence) {
        currentTrajectorySequence = trajectorySequence;
        currentSegmentStartTime = clock.seconds();
        currentSegmentIndex = 0;
        lastSegmentIndex = -1;
    }

    public @Nullable
    DriveSignal update(Pose2d poseEstimate, Pose2d poseVelocity) {
        Pose2d targetPose = null;
        DriveSignal driveSignal = null;

        TelemetryPacket packet = new TelemetryPacket();
        Canvas fieldOverlay = packet.fieldOverlay();

        SequenceSegment currentSegment = null;

        if (currentTrajectorySequence != null) {
            if (currentSegmentIndex >= currentTrajectorySequence.size()) {
                for (TrajectoryMarker marker : remainingMarkers) {
                    marker.getCallback().onMarkerReached();
                }

                remainingMarkers.clear();

                currentTrajectorySequence = null;
            }

            if (currentTrajectorySequence == null)
                return new DriveSignal();

            double now = clock.seconds();
            boolean isNewTransition = currentSegmentIndex != lastSegmentIndex;

            currentSegment = currentTrajectorySequence.get(currentSegmentIndex);

            if (isNewTransition) {
                currentSegmentStartTime = now;
                lastSegmentIndex = currentSegmentIndex;

                for (TrajectoryMarker marker : remainingMarkers) {
                    marker.getCallback().onMarkerReached();
                }

                remainingMarkers.clear();

                remainingMarkers.addAll(currentSegment.getMarkers());
                Collections.sort(remainingMarkers, (t1, t2) -> Double.compare(t1.getTime(), t2.getTime()));
            }

            double deltaTime = now - currentSegmentStartTime;

            if (currentSegment instanceof TrajectorySegment) {
                Trajectory currentTrajectory = ((TrajectorySegment) currentSegment).getTrajectory();

                if (isNewTransition)
                    follower.followTrajectory(currentTrajectory);

                if (!follower.isFollowing()) {
                    currentSegmentIndex++;

                    driveSignal = new DriveSignal();
                } else {
                    driveSignal = follower.update(poseEstimate, poseVelocity);
                    lastPoseError = follower.getLastError();
                }

                targetPose = currentTrajectory.get(deltaTime);
            } else if (currentSegment instanceof TurnSegment) {
                MotionState targetState = ((TurnSegment) currentSegment).getMotionProfile().get(deltaTime);

                turnController.setTargetPosition(targetState.getX());

                double correction = turnController.update(poseEstimate.getHeading());

                double targetOmega = targetState.getV();
                double targetAlpha = targetState.getA();

                lastPoseError = new Pose2d(0, 0, turnController.getLastError());

                Pose2d startPose = currentSegment.getStartPose();
                targetPose = startPose.copy(startPose.getX(), startPose.getY(), targetState.getX());

                driveSignal = new DriveSignal(
                        new Pose2d(0, 0, targetOmega + correction),
                        new Pose2d(0, 0, targetAlpha)
                );

                if (deltaTime >= currentSegment.getDuration()) {
                    currentSegmentIndex++;
                    driveSignal = new DriveSignal();
                }
            } else if (currentSegment instanceof WaitSegment) {
                lastPoseError = new Pose2d();

                targetPose = currentSegment.getStartPose();
                driveSignal = new DriveSignal();

                if (deltaTime >= currentSegment.getDuration()) {
                    currentSegmentIndex++;
                }
            }

            while (remainingMarkers.size() > 0 && deltaTime > remainingMarkers.get(0).getTime()) {
                remainingMarkers.get(0).getCallback().onMarkerReached();
                remainingMarkers.remove(0);
            }
        }

        poseHistory.add(poseEstimate);

        if (POSE_HISTORY_LIMIT > -1 && poseHistory.size() > POSE_HISTORY_LIMIT) {
            poseHistory.removeFirst();
        }

        final double NOMINAL_VOLTAGE = 12.0;
        double voltage = voltageSensor.getVoltage();
        if (driveSignal != null && !DriveConstants.RUN_USING_ENCODER) {
            driveSignal = new DriveSignal(
                    driveSignal.getVel().times(NOMINAL_VOLTAGE / voltage),
                    driveSignal.getAccel().times(NOMINAL_VOLTAGE / voltage)
            );
        }

        if (targetPose != null) {
            LogFiles.record(
                    targetPose, poseEstimate, voltage,
                    lastDriveEncPositions, lastDriveEncVels, lastTrackingEncPositions, lastTrackingEncVels
            );
        }

        packet.put("x", poseEstimate.getX());
        packet.put("y", poseEstimate.getY());
        packet.put("heading (deg)", Math.toDegrees(poseEstimate.getHeading()));

        packet.put("xError", getLastPoseError().getX());
        packet.put("yError", getLastPoseError().getY());
        packet.put("headingError (deg)", Math.toDegrees(getLastPoseError().getHeading()));

        draw(fieldOverlay, currentTrajectorySequence, currentSegment, targetPose, poseEstimate);

        dashboard.sendTelemetryPacket(packet);

        return driveSignal;
    }

    private void draw(
            Canvas fieldOverlay,
            TrajectorySequence sequence, SequenceSegment currentSegment,
            Pose2d targetPose, Pose2d poseEstimate
    ) {
        if (sequence != null) {
            for (int i = 0; i < sequence.size(); i++) {
                SequenceSegment segment = sequence.get(i);

                if (segment instanceof TrajectorySegment) {
                    fieldOverlay.setStrokeWidth(1);
                    fieldOverlay.setStroke(COLOR_INACTIVE_TRAJECTORY);

                    DashboardUtil.drawSampledPath(fieldOverlay, ((TrajectorySegment) segment).getTrajectory().getPath());
                } else if (segment instanceof TurnSegment) {
                    Pose2d pose = segment.getStartPose();

                    fieldOverlay.setFill(COLOR_INACTIVE_TURN);
                    fieldOverlay.fillCircle(pose.getX(), pose.getY(), 2);
                } else if (segment instanceof WaitSegment) {
                    Pose2d pose = segment.getStartPose();

                    fieldOverlay.setStrokeWidth(1);
                    fieldOverlay.setStroke(COLOR_INACTIVE_WAIT);
                    fieldOverlay.strokeCircle(pose.getX(), pose.getY(), 3);
                }
            }
        }

        if (currentSegment != null) {
            if (currentSegment instanceof TrajectorySegment) {
                Trajectory currentTrajectory = ((TrajectorySegment) currentSegment).getTrajectory();

                fieldOverlay.setStrokeWidth(1);
                fieldOverlay.setStroke(COLOR_ACTIVE_TRAJECTORY);

                DashboardUtil.drawSampledPath(fieldOverlay, currentTrajectory.getPath());
            } else if (currentSegment instanceof TurnSegment) {
                Pose2d pose = currentSegment.getStartPose();

                fieldOverlay.setFill(COLOR_ACTIVE_TURN);
                fieldOverlay.fillCircle(pose.getX(), pose.getY(), 3);
            } else if (currentSegment instanceof WaitSegment) {
                Pose2d pose = currentSegment.getStartPose();

                fieldOverlay.setStrokeWidth(1);
                fieldOverlay.setStroke(COLOR_ACTIVE_WAIT);
                fieldOverlay.strokeCircle(pose.getX(), pose.getY(), 3);
            }
        }

        if (targetPose != null) {
            fieldOverlay.setStrokeWidth(1);
            fieldOverlay.setStroke("#4CAF50");
            DashboardUtil.drawRobot(fieldOverlay, targetPose);
        }

        fieldOverlay.setStroke("#3F51B5");
        DashboardUtil.drawPoseHistory(fieldOverlay, poseHistory);

        fieldOverlay.setStroke("#3F51B5");
        DashboardUtil.drawRobot(fieldOverlay, poseEstimate);
    }

    public Pose2d getLastPoseError() {
        return lastPoseError;
    }

    public boolean isBusy() {
        return currentTrajectorySequence != null;
    }
}

File: TeamCode/src/main/kotlin/org/firstinspires/ftc/teamcode/trajectorysequence/TrajectorySequenceRunner.kt
package org.firstinspires.ftc.teamcode.trajectorysequence

import com.acmerobotics.dashboard.FtcDashboard
import com.acmerobotics.dashboard.canvas.Canvas
import com.acmerobotics.dashboard.config.Config
import com.acmerobotics.dashboard.telemetry.TelemetryPacket
import com.acmerobotics.roadrunner.control.PIDCoefficients
import com.acmerobotics.roadrunner.control.PIDFController
import com.acmerobotics.roadrunner.drive.DriveSignal
import com.acmerobotics.roadrunner.followers.TrajectoryFollower
import com.acmerobotics.roadrunner.geometry.Pose2d
import com.acmerobotics.roadrunner.profile.MotionState
import com.acmerobotics.roadrunner.trajectory.Trajectory
import com.acmerobotics.roadrunner.trajectory.TrajectoryMarker
import com.acmerobotics.roadrunner.util.NanoClock
import com.qualcomm.robotcore.hardware.VoltageSensor
import org.firstinspires.ftc.teamcode.config.DriveConstants
import org.firstinspires.ftc.teamcode.trajectorysequence.sequencesegment.SequenceSegment
import org.firstinspires.ftc.teamcode.trajectorysequence.sequencesegment.TrajectorySegment
import org.firstinspires.ftc.teamcode.trajectorysequence.sequencesegment.TurnSegment
import org.firstinspires.ftc.teamcode.trajectorysequence.sequencesegment.WaitSegment
import org.firstinspires.ftc.teamcode.util.DashboardUtil
import org.firstinspires.ftc.teamcode.util.LogFiles

@Config
class TrajectorySequenceRunner(
    private val follower: TrajectoryFollower,
    headingPIDCoefficients: PIDCoefficients,
    voltageSensor: VoltageSensor,
    lastDriveEncPositions: List<Int>,
    lastDriveEncVels: List<Int>,
    lastTrackingEncPositions: List<Int>,
    lastTrackingEncVels: List<Int>
) {

    private val turnController: PIDFController = PIDFController(headingPIDCoefficients)

    private val clock: NanoClock

    private var currentTrajectorySequence: TrajectorySequence? = null

    private var currentSegmentStartTime = 0.0
    private var currentSegmentIndex = 0
    private var lastSegmentIndex = 0

    private var lastPoseError: Pose2d = Pose2d()

    private var remainingMarkers: MutableList<TrajectoryMarker> = java.util.ArrayList<TrajectoryMarker>()

    private val dashboard: FtcDashboard
    private val poseHistory: java.util.LinkedList<Pose2d> = java.util.LinkedList<Pose2d>()

    private val voltageSensor: VoltageSensor

    private val lastDriveEncPositions: List<Int>
    private val lastDriveEncVels: List<Int>
    private val lastTrackingEncPositions: List<Int>
    private val lastTrackingEncVels: List<Int>

    init {
        turnController.setInputBounds(0.0, 2 * Math.PI)

        this.voltageSensor = voltageSensor

        this.lastDriveEncPositions = lastDriveEncPositions
        this.lastDriveEncVels = lastDriveEncVels
        this.lastTrackingEncPositions = lastTrackingEncPositions
        this.lastTrackingEncVels = lastTrackingEncVels

        clock = NanoClock.system()

        dashboard = FtcDashboard.getInstance()
        dashboard.telemetryTransmissionInterval = 25
    }

    fun followTrajectorySequenceAsync(trajectorySequence: TrajectorySequence) {
        currentTrajectorySequence = trajectorySequence
        currentSegmentStartTime = clock.seconds()
        currentSegmentIndex = 0
        lastSegmentIndex = -1
    }

    fun update(poseEstimate: Pose2d, poseVelocity: Pose2d?): DriveSignal? {
        var targetPose: Pose2d? = null
        var driveSignal: DriveSignal? = null

        val packet = TelemetryPacket()
        val fieldOverlay = packet.fieldOverlay()

        if (currentTrajectorySequence == null) {
            return null
        }

        if (currentSegmentIndex >= currentTrajectorySequence?.size()!!) {
            for (marker in remainingMarkers) {
                marker.callback.onMarkerReached()
            }

            remainingMarkers.clear()

            currentTrajectorySequence = null

            return null
        }

        val now: Double = clock.seconds()
        val isNewTransition = currentSegmentIndex != lastSegmentIndex

        val currentSegment = currentTrajectorySequence!!.get(currentSegmentIndex)

        if (isNewTransition) {
            currentSegmentStartTime = now
            lastSegmentIndex = currentSegmentIndex

            for (marker in remainingMarkers) {
                marker.callback.onMarkerReached()
            }

            remainingMarkers.clear()

            remainingMarkers.addAll(currentSegment.markers)
            remainingMarkers.sortWith { t1, t2 -> t1.time.compareTo(t2.time) }
        }

        val deltaTime = now - currentSegmentStartTime

        if (currentSegment is TrajectorySegment) {
            val currentTrajectory: Trajectory = currentSegment.trajectory

            if (isNewTransition) follower.followTrajectory(currentTrajectory)

            if (!follower.isFollowing()) {
                currentSegmentIndex++

                driveSignal = DriveSignal()
            } else {
                driveSignal = follower.update(poseEstimate, poseVelocity)
                lastPoseError = follower.lastError
            }

            targetPose = currentTrajectory[deltaTime]
        } else if (currentSegment is TurnSegment) {
            val targetState: MotionState = currentSegment.motionProfile[deltaTime]

            turnController.targetPosition = targetState.x

            val correction: Double = turnController.update(poseEstimate.heading)

            val targetOmega: Double = targetState.v
            val targetAlpha: Double = targetState.a

            lastPoseError = Pose2d(0.0, 0.0, turnController.lastError)

            val startPose: Pose2d = currentSegment.startPose
            targetPose = startPose.copy(startPose.x, startPose.y, targetState.x)

            driveSignal = DriveSignal(
                Pose2d(0.0, 0.0, targetOmega + correction),
                Pose2d(0.0, 0.0, targetAlpha)
            )

            if (deltaTime >= currentSegment.duration) {
                currentSegmentIndex++
                driveSignal = DriveSignal()
            }
        } else if (currentSegment is WaitSegment) {
            lastPoseError = Pose2d()

            targetPose = currentSegment.startPose
            driveSignal = DriveSignal()

            if (deltaTime >= currentSegment.duration) {
                currentSegmentIndex++
            }
        }

        while (remainingMarkers.size > 0 && deltaTime > remainingMarkers[0].time) {
            remainingMarkers[0].callback.onMarkerReached()
            remainingMarkers.removeAt(0)
        }

        poseHistory.add(poseEstimate)

        if (POSE_HISTORY_LIMIT > -1 && poseHistory.size > POSE_HISTORY_LIMIT) {
            poseHistory.removeFirst()
        }

        val nominalVoltage = 12.0
        val voltage: Double = voltageSensor.voltage
        if (driveSignal != null && !DriveConstants.RUN_USING_ENCODER) {
            driveSignal = DriveSignal(
                driveSignal.vel.times(nominalVoltage / voltage),
                driveSignal.accel.times(nominalVoltage / voltage)
            )
        }

        if (targetPose != null) {
            LogFiles.record(
                targetPose, poseEstimate, voltage,
                lastDriveEncPositions, lastDriveEncVels, lastTrackingEncPositions, lastTrackingEncVels
            )
        }

        packet.put("x", poseEstimate.x)
        packet.put("y", poseEstimate.y)
        packet.put("heading (deg)", Math.toDegrees(poseEstimate.heading))

        packet.put("xError", getLastPoseError().x)
        packet.put("yError", getLastPoseError().y)
        packet.put("headingError (deg)", Math.toDegrees(getLastPoseError().heading))

        draw(fieldOverlay, currentTrajectorySequence, currentSegment, targetPose, poseEstimate)

        dashboard.sendTelemetryPacket(packet)

        return driveSignal
    }

    private fun draw(
        fieldOverlay: Canvas,
        sequence: TrajectorySequence?, currentSegment: SequenceSegment?,
        targetPose: Pose2d?, poseEstimate: Pose2d
    ) {
        if (sequence != null) {
            for (i in 0 until sequence.size()) {
                when (val segment: SequenceSegment = sequence.get(i)) {
                    is TrajectorySegment -> {
                        fieldOverlay.setStrokeWidth(1)
                        fieldOverlay.setStroke(COLOR_INACTIVE_TRAJECTORY)

                        DashboardUtil.drawSampledPath(
                            fieldOverlay,
                            segment.trajectory.path
                        )
                    }

                    is TurnSegment -> {
                        val pose: Pose2d = segment.startPose

                        fieldOverlay.setFill(COLOR_INACTIVE_TURN)
                        fieldOverlay.fillCircle(pose.x, pose.y, 2.0)
                    }

                    is WaitSegment -> {
                        val pose: Pose2d = segment.startPose

                        fieldOverlay.setStrokeWidth(1)
                        fieldOverlay.setStroke(COLOR_INACTIVE_WAIT)
                        fieldOverlay.strokeCircle(pose.x, pose.y, 3.0)
                    }
                }
            }
        }

        if (currentSegment != null) {
            when (currentSegment) {
                is TrajectorySegment -> {
                    val currentTrajectory: Trajectory = currentSegment.trajectory

                    fieldOverlay.setStrokeWidth(1)
                    fieldOverlay.setStroke(COLOR_ACTIVE_TRAJECTORY)

                    DashboardUtil.drawSampledPath(fieldOverlay, currentTrajectory.path)
                }

                is TurnSegment -> {
                    val pose: Pose2d = currentSegment.startPose

                    fieldOverlay.setFill(COLOR_ACTIVE_TURN)
                    fieldOverlay.fillCircle(pose.x, pose.y, 3.0)
                }

                is WaitSegment -> {
                    val pose: Pose2d = currentSegment.startPose

                    fieldOverlay.setStrokeWidth(1)
                    fieldOverlay.setStroke(COLOR_ACTIVE_WAIT)
                    fieldOverlay.strokeCircle(pose.x, pose.y, 3.0)
                }
            }
        }

        if (targetPose != null) {
            fieldOverlay.setStrokeWidth(1)
            fieldOverlay.setStroke("#4CAF50")
            DashboardUtil.drawRobot(fieldOverlay, targetPose)
        }

        fieldOverlay.setStroke("#3F51B5")
        DashboardUtil.drawPoseHistory(fieldOverlay, poseHistory)

        fieldOverlay.setStroke("#3F51B5")
        DashboardUtil.drawRobot(fieldOverlay, poseEstimate)
    }

    fun getLastPoseError(): Pose2d {
        return lastPoseError
    }

    val isBusy: Boolean
        get() = currentTrajectorySequence != null

    companion object {
        var COLOR_INACTIVE_TRAJECTORY: String = "#4caf507a"
        var COLOR_INACTIVE_TURN: String = "#7c4dff7a"
        var COLOR_INACTIVE_WAIT: String = "#dd2c007a"

        var COLOR_ACTIVE_TRAJECTORY: String = "#4CAF50"
        var COLOR_ACTIVE_TURN: String = "#7c4dff"
        var COLOR_ACTIVE_WAIT: String = "#dd2c00"

        var POSE_HISTORY_LIMIT: Int = 100
    }
}